Java LoC metric

dburriss opened the pull request:
This PR is for adding the LoC metric discussed in issue #359.

marco-c commented:
@dburriss still interested in finishing this up?

dburriss commented:
Yes. Just not a lot of time at the moment. Hopefully be able to come back to this in a week or 2.

dburriss commented:
@marco-c @Luni-4 this is up-to-date with master and ready for review.

calixteman commented on src/getter.rs:
```diff
@@ -509,4 +509,17 @@ impl Getter for CppCode {
+        let typ = node.object().kind_id();
+        match typ.into() {
+            MethodDeclaration | LambdaExpression => SpaceKind::Function,
+            Class | ClassDeclaration => SpaceKind::Class,
```
Did you intentionally miss interface:
https://github.com/tree-sitter/tree-sitter-java/blob/a24ae7d16de3517bff243a87d087d0b4877a65c5/grammar.js#L632
?
If yes, please add a comment to explain why.
In looking at the above link I can see that there is a `module` concept (I don't know what is: I didn't write Java code for years), so same question.

dburriss replied:
Good catch. Since this is my first time working in Rust, this repo, and tree-sitter, rather than copying other implementations I just added grammar to make tests pass. 
I will take a scan through to see if any other things I missed.
It has been a decade or so since I have looked at Java... I do not remember ever using `module` :D I learned something today.

dburriss replied:
For the SLOC off by 1 issue the only match I need is `Program => SpaceKind::Unit`. The space_kind for the rest can be handled in the implementations that need them. Otherwise no test coverage. Does that make sense to you @calixteman ?

calixteman replied:
What's the SLOC off by 1 issue ?

dburriss replied:
Well, I left this note for myself while documenting my changes. And no I don't generally make my notes as if I am talking to myself but I had initially thought to add to the documentation as I went.

> As you implement tests, you may find yourself running into an off by 1 error for SLOC count. 
> For most code we expect the `SpaceKind` to be `SpaceKind::Unit` but the default is `SpaceKind::Unknown`.
> To solve this you need to implement the `Getter` trait in [src/getter.rs](https://github.com/mozilla/rust-code-analysis/blob/master/src/getter.rs). Specifically, the `get_space_kind` function.

I think it is because any java snippet of code gets a `Program` leaf. I guess `UNKNOWN` gets counted and `UNIT` does not? Or maybe the other way around? It was a few months ago that I wrote it...

dburriss replied:
@calixteman does this one make sense? I imagine the rest will be added for other metrics as needed. Possibly needed for other metrics counting functions? I have not looked at anything but LoC and the Java Cyclomatic PR.

calixteman commented on src/metrics/loc.rs:
```diff
@@ -500,9 +500,59 @@ impl Loc for CppCode {
 
+impl fmt::Display for Java {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
```
I don't see similar code for other languages, is it a debug leftover ?
If it's useful in general, I guess it can be added for all languages in a follow-up PR.

dburriss replied:
It indeed is from debugging. Ok to leave in then? Should I create an issue to track adding it to the others? Guss can discuss there and if we decide no, then I will remove this.

calixteman replied:
I would prefer having that in an other PR for all the different languages.

dburriss replied:
Will do. I will pull it out of this PR once the other items in this are sorted out. I have not found a better way of seeing what is going on.

calixteman commented on src/metrics/loc.rs:
```diff
@@ -500,9 +500,59 @@ impl Loc for CppCode {
+            Comment => {
+                add_cloc_lines(stats, start, end);
+            }
+            AssertStatement
```
Maybe you missed `InstanceofExpression` and few others, have a look on:
https://github.com/tree-sitter/tree-sitter-java/blob/a24ae7d16de3517bff243a87d087d0b4877a65c5/grammar.js#L161-L170

calixteman replied:
@dburriss ^^

dburriss replied:
@calixteman  I should have mentioned it explicitly. I did add these in but it completely breaks the counts. I will record some examples. Maybe we can get to the same counts with fewer statements and more expressions.

dburriss replied:
For ease of discussion, here is the result of parsing that:
```
LocalVariableDeclaration
TypeIdentifier
VariableDeclarator
Identifier
EQ
TernaryExpression
BinaryExpression
ParenthesizedExpression
LPAREN
BinaryExpression
Identifier
PERCENT
DecimalIntegerLiteral
RPAREN
EQEQ
DecimalIntegerLiteral
QMARK
StringLiteral
COLON
StringLiteral
SEMI
Comment
```

calixteman replied:
I re-read the sloc & lloc definitions on wikipedia:
https://en.wikipedia.org/wiki/Source_lines_of_code#Measurement_methods
and in lloc we should have only statements.
I wonder why I talked about InstanceofExpression... maybe because there was an Expression here in the 1st version of your patch.... or maybe I was a bit tired... Anyway sorry for that.
For example with js we only have statements for logical_lines:
https://github.com/mozilla/rust-code-analysis/blob/d488b6eb5b4de522700a75a84602d404a5f64de1/src/metrics/loc.rs#L343-L349
But for rust code:
https://github.com/mozilla/rust-code-analysis/blob/d488b6eb5b4de522700a75a84602d404a5f64de1/src/metrics/loc.rs#L424-L436
the term `Expression` is used in place of `Statement`: it's a bit confusing.
So it really depends on the way the grammar is built itself.

Anyway, for Java we've a a grammar definition for statement:
https://github.com/tree-sitter/tree-sitter-java/blob/a24ae7d16de3517bff243a87d087d0b4877a65c5/grammar.js#L415
so I guess we can basically re-use what we've here and nothing else (like for js, python stuff).

dburriss replied:
Yeah definitely depends on the language. I guess Rust takes some FP cues so expressions are preferred. Statements were the original push as discussed with Luni-4 https://github.com/mozilla/rust-code-analysis/issues/359#issuecomment-952631456. Sorrry, I should have thought to link to it sooner.
I'll revert back in the next day or 2. I learned a bit about Java. Have not touched it since version 6 so if nothing else I learned a few things about how it does lambdas. I must admit, I like it even less than I like its generics but hey. Each to their own.

dburriss replied:
All expressions are removed now.

calixteman commented on src/metrics/loc.rs:
```diff
@@ -1428,4 +1477,284 @@ mod tests {
+            
+            class HelloWorldApp {
+              public void main(String[] args) {
+                String message = args.length == 0 ? \"Hello empty world\" : \"Hello world\"; // +2 lloc : 1 var assignment + binary exp
```
yes it does.

calixteman replied:
So after reading lloc doc on wikipedia: we should have 1 here: for all the assignment, which should be an ExpressionStatement.

dburriss replied:
@calixteman I missed this comment but I had a similar thought. I went with `ExpressionStatement` instead of method invocation and a few other things like handling `i++;`.
This specific line doesn't parse as an expression statement but does count as 1 LLOC as it is a `LocalVariableDeclaration`.

dburriss commented:
@calixteman A quick summary of the latest changes in commit "Remove expressions from count".

- I removed ALL Expression matches
- IMPORTANT: with the removal of `BinaryExpression` `UpdateExpression` means the numbers for a `for` would drop for 3 to 2. I will call this out with a comment specifically.
- updated the expectations in the tests from the expression changes. The major change in count is around ternary as is expected. 
- added some tests and changed a few others to cover `try`, `switch`, `do..while`, `throw`, `break`, and `continue`.

A reminder: I will remove the debug stuff once we are happy with the implementation.

calixteman commented on src/metrics/loc.rs:
```diff
@@ -500,9 +500,55 @@ impl Loc for CppCode {
+            }
+            ForStatement => {
+                // handle binary check + increment
+                stats.logical_lines += 2;
```
mmh... on wikipedia:
https://en.wikipedia.org/wiki/Source_lines_of_code
we have 1 for the `for` statement.
Tbh, from our pov, we don't care that much: we use these numbers in some machine learning stuff so I guess it doesn't make any difference at the end.
So I think we must be consistent in the way to compute those numbers across the different languages.
@marco-c, @Luni-4 do we have, in the scientific literature, some clear definitions ?

dburriss replied:
@Luni-4 Thanks. A good suggestion to link in the code. I will add that.
I did read the article when you linked to it in [this discussion](https://github.com/mozilla/rust-code-analysis/issues/359#issuecomment-952631456) and we agreed then that the classic `for` loop had a LLOC of 3, in the same comment.

From that Oracle article:

> Assignment expressions
> Any use of ++ or --
> Method invocations
> Object creation expressions

Not counting the binary expression but counting the `for` loop itself (like is done with `while` and `do...while`), still puts us at a count of 3 if we include the `++` as the article suggests. This all depends if we decide that what is inside a `for` loop is subject to those rules.

[This article from Oracle](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/for.html) refers to the initializer, condition, and incrementer as *expressions*. So since it is specific to the case we are talking about, my suggestion is we do as @Luni-4 suggests and make any `for` count as ` LLOC. That simplifies the code AND seems to be the most straightforward way to interpret those articles taken together.

So. `for` = 1 LLOC. Agreed?

Luni-4 replied:
Since Java considers a `for` as a single statement and its internals as expressions, I agree with the `for` = 1 LLOC line. Let's link both the articles in the code as comments though, so we know we have decided one way instead of the other

calixteman replied:
yep, agreed too.

dburriss commented:
Ok in this latest push I have incorporated the changes from the comments from @Luni-4 (hopefully @calixteman you agree with the arguments)

- `for` statement is ` LLOC and so expressions inside are ignored
- commented with links to the Oracle articles on why we only count statements, including the one about `for` statements
- removed the debug lines etc.
- rebased the cyclomatic complexity and other changes from the past week or 2

Hopefully, it is ready to go.

dburriss commented:
@calixteman would you have a chance to review this before the weekend? That way I can make any requested changes over the weekend or move on to another.

marco-c commented:
@dburriss Calixte is off this week, perhaps you could start looking at C#?

marco-c commented:
Calixte was off this week too unfortunately, he's back today :)
BTW, you could join our Matrix room on https://chat.mozilla.org/#/room/#rust-code-analysis:mozilla.org.

dburriss commented:
Rebased in recent merges and fixed errors from recent `Stats` changes.

calixteman commented on src/getter.rs:
```diff
@@ -509,4 +509,14 @@ impl Getter for CppCode {
+
+        let typ = node.object().kind_id();
+        match typ.into() {
+            Program => SpaceKind::Unit,
```
My definition of space is: something which can have its own metrics.
So for example, you can have a class A with an inner class B each of them could be a space: this way we'll have some metrics for both of them:
 - B has its own metrics
 - A has its own metrics merged with the B ones
And at the end in the final output you'll have the granularity the user wants.
So not having more spaces here means that we don't have any granularity.

dburriss replied:
Interesting. I have taken a shot at it based on the other implementations and that explanation. Other than that not having `Program => SpaceKind::Unit` causes the line count to be off (javascript and typescript have this match too), I am uncertain how to actually test this in the context of LoC?

dburriss replied:
@calixteman I think this covers what can contain other code in Java. As I said in the above comment, I am uncertain how to test this in the context of LoC.

```rust
ClassDeclaration => SpaceKind::Class,
MethodDeclaration | LambdaExpression => SpaceKind::Function,
PackageDeclaration | ModuleDeclaration => SpaceKind::Namespace,
Program => SpaceKind::Unit
```

calixteman replied:
If you want to see space in actions you could just run `cargo run -p rust-code-analysis-cli -- -m -p foo.java` and have a look on the output, with foo.java:
```java
class Outer {
  String x;
 
  class Inner {
    String y;
    int bar() {
       return 12;
    }
  }
}
```

If you return unknown for any kind of space, you'll get some `unknown` in the output.
Maybe the "off by 1" issue is related to:
https://github.com/mozilla/rust-code-analysis/blob/master/src/metrics/loc.rs#L40

dburriss replied:
Ah cool! Makes complete sense. And yes that line of code was what led me to the fix of `Unit` for `Program` for the "off by 1" issue. It was in November last year when I started on this (and learning Rust) so the memories are a little hazy.

Adding a lambda expression and a package to the java file didn't seem to add anything to the structure. Not sure if I should leave them in, conceptually it seems correct to me as is. The code on this comment is not showing as outdated but I do think the kind is working as you wanted. See image:

![image](https://user-images.githubusercontent.com/5071495/159553284-c9d1a861-165e-4bf5-8910-db44963d5e07.png)

calixteman commented:
Sorry for the long delay: I've been to be off in the last weeks (life....).
Overall it looks good, just need to address the space thing issue and then I guess it'll be mergeable.

calixteman commented on src/getter.rs:
```diff
@@ -509,4 +509,17 @@ impl Getter for CppCode {
+
+        let typ = node.object().kind_id();
+        match typ.into() {
+            ClassDeclaration => SpaceKind::Class,
```
Maybe you could add InterfaceDeclaration here too.
Maybe it's worthing adding a SpaceKind::Interface because an interface contributes to loc for example but it expected to have other metrics equal to 0 since it just some functions declarations.
So having interface in the output will help the user the understand this numbers and maybe (probably ?) skip this data since they aren't really useful (but maybe they're).

dburriss replied:
I mapped `InterfaceDeclaration` as a `SpaceKind::Class` and it didn't show up at all in the hierarchy.

With AND without gave:

unit: foo.java
- function: print
- class: A6
-- function: print
-- function: main 

for the following code file:

```java
package com.java;

interface printable{  
  void print();  
}  

class A6 implements printable{  
  public void print(){System.out.println("Hello");}  
    
  public static void main(String args[]){  
    A6 obj = new A6();  
    obj.print();  
  }  
} 
```

Adding a `SpaceKind::Interface` interface could be interesting if it indeed showed up in the hierarchy. I would expect:

unit: foo.java
- interface:  printable
-- function: print
- class: A6
-- function: print
-- function: main 

Although as you said, not sure how many metrics interfaces actually contribute. Evaluating with "Lay of least surprise" I do think the `SpaceKind::Interface` makes sense but that should be a separate PR I would think?

calixteman reviewed (approved):
It looks good to me.
Thanks a lot for doing that.
